### DISCUSSION TITLE
useSpring vs useTransition

### DISCUSSION THREAD

**Discussion by LoyalPotato** [pmndrs/react-spring#2012] [Support] [answered]
Hello,

My question is when do I use one or the other.
In the `useSpring` page there's an example for a [file tree](https://codesandbox.io/s/github/pmndrs/react-spring/tree/master/demo/src/sandboxes/tree) and it's using a state flag to change the height. Wouldn't `useTransition` be best for this?

**Comment by joshuaellis** [comment, marked as answer]
>My question is when do I use one or the other.
In the useSpring page there's an example for a [file tree](https://codesandbox.io/s/github/pmndrs/react-spring/tree/master/demo/src/sandboxes/tree) and it's using a state flag to change the height. Wouldn't useTransition be best for this?

I think it depends what you want to do, you could use `useTransition` you could use `useSpring`, I don't think you get any benefit from removing the items from the DOM. 

I normally start with `useSpring` and use `useTransition` sparingly.

**Reply by LoyalPotato** [reply]
Could you clarify the part about removing from the DOM?

**Reply by joshuaellis** [reply]
useTransition allows you to render items to the dom depending on the data so say you have a list:

```js
[1,2,3,4]
```

when you render those items you'll make DOM nodes for each one. and when the list changes – `[1,2,3]` no. 4 will be removed. But sometimes e.g. a navigation list even though the items aren't visible they want to still be in the DOM because you want it to still be accessible by keyboard navigation. I would suspect this is the case for a tree-view.
